tests/teststun.nim

```diff
@@ -0,0 +1,87 @@
+# Nim-WebRTC
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+{.used.}
+
+import options, strutils
+import bearssl
+import ../webrtc/udp_connection
+import ../webrtc/stun/stun_connection
+import ../webrtc/stun/stun_message
+import ../webrtc/stun/stun_attributes
+import ../webrtc/stun/stun_utils
+import ./asyncunit
+
+proc newRng(): ref HmacDrbgContext =
+  HmacDrbgContext.new()
+
+suite "Stun message encoding/decoding":
+  test "Get BindingRequest + encode & decode with a set username":
+    let
+      udpConn = UdpConn.init(AnyAddress)
+      conn = StunConn.init(udpConn, TransportAddress(AnyAddress), true, newRng())
+      msg = conn.getBindingRequest(username = "Test")
+      encoded = msg.encode(msg.getAttribute(AttrUsername))
+      decoded = StunMessage.decode(encoded)
+
+    check:
+      msg.msgType == decoded.msgType
+      msg.transactionId == decoded.transactionId
+      msg.getAttribute(AttrUsername) == decoded.getAttribute(AttrUsername)
+      msg.getAttribute(AttrICEControlling).isSome()
+      msg.getAttribute(AttrICEControlled).isNone()
+      Priority.decode(msg.getAttribute(AttrPriority).get()).priority == 0x7effffff
+      # encoding adds Fingerprint as attributes
+      msg.attributes.len() == decoded.attributes.len() - 1
+    conn.close()
+
+  test "Get BindingRequest + encode & decode with a random username":
+    let
+      udpConn = UdpConn.init(AnyAddress)
+      conn = StunConn.init(udpConn, TransportAddress(AnyAddress), false, newRng())
+      msg = conn.getBindingRequest()
+      encoded = msg.encode(msg.getAttribute(AttrUsername))
+      decoded = StunMessage.decode(encoded)
+
+    check:
+      msg.msgType == decoded.msgType and msg.msgType == StunBindingRequest
+      msg.transactionId == decoded.transactionId
+      msg.getAttribute(AttrUsername) == decoded.getAttribute(AttrUsername)
+      msg.getAttribute(AttrICEControlling).isNone()
+      msg.getAttribute(AttrICEControlled).isSome()
+      Priority.decode(msg.getAttribute(AttrPriority).get()).priority == 0x7effffff
+      # encoding adds Fingerprint as attributes
+      msg.attributes.len() == decoded.attributes.len() - 2
```

Comment:
Why it's -2 here and not -1 like above? There's some repetition too in the checks, maybe we can create a proc for that? Here there's `msg.msgType == StunBindingRequest`, but above there isn't. This kind of discrepancy tends to happen where there's duplicate code doing the same thing and a change isn't applied to all places.

Reply:
Because in this case, the encoding added both Fingerprint and Message-integrity.
I updated the comment.
And I tried to implement a function to compare msg and decoded to avoid duplicating the code. It crashed two or three times and I couldn't understand why. So I give up and revert back to this.

Reply:
But should it crash? Might be a sign there's something wrong.

Reply:
From what I saw, it looked more like a `me doing dumb things with unittest` than `something don't work with the encoding/decoding`

Reply:
Is it related to unittest? I thought about a proc like that https://github.com/vacp2p/nim-libp2p/pull/1073/files#diff-f8c32e08e834a2cd25291ef960641e67cd42b1b30ab9c2dbceab16099ef8b6b1R33

Reply:
Oh right, it could work indeed.

Reply:
I reworked the encoding/decoding tests.